text/0064-refer-to-types-in-empty-namespace.md

```diff
@@ -0,0 +1,159 @@
+* regardless of whether `id` is the name of a common type or an entity type, in
+either the empty and/or active namespace
+* in both the human (RFC 24) and JSON schema syntaxes
+
+This does not apply:
+* to primitive and extension types. From the user's perspective, those live in
+the `__cedar` namespace and not in the empty namespace. `::String` refers to an
+entity or common type `String` defined in the empty namespace, while
+`__cedar::String` refers to the builtin primitive type `String`.
+* to "names" like `::A::B` -- that syntax remains an error. `A::B` is already an
+unambiguous, fully-qualified name, and will never have any namespace implicitly
+prepended.
+
+### Change 2: Implicitly falling back on the empty namespace
+
+When the schema refers to an unqualified typename (e.g, `id`) which doesn't
+exist in the active namespace, we should fall back to looking for it in the
+empty namespace.
+Only if it exists in neither the active nor empty namespace should this be an
+error.
+
+E.g., in the example at the top of this RFC, consider the type of the `id`
+attribute of `Demo::User`.
+Since there is no `id` type defined in namespace `Demo`, this is an error today.
+With this change, when we see there is no `Demo::id`, we implicitly look for
+`id` in the empty namespace, and resolving `id` to be the common type `::id`
+defined to be `String` in the empty namespace.
+The user could, of course, write `::id` directly, but with this change, they
+do not have to, in the common case where there is no collision (where `id` is
+not actually defined in both the `Demo` and empty namespaces).
+
+This applies to both entity and common types defined in the empty namespace.
+
+This applies to both the human (RFC 24) and JSON schema syntaxes.
+
+This is reminiscent of, but not dependent on, the existing similar behavior for
+primtiive and extension types, where we first check for an entity or common type
+`String` in the active namespace before falling back on resolving to
+`__cedar::String`.
+
+## Drawbacks
+
+* The schema implementation becomes more complex, as we need more complex
+methods for typename resolution, and also slightly more complexity in the
+schema parser, as `::id` wasn't previously legal syntax (`::` previously was
+only legal between two identifiers).
+
+This is not a breaking change; all existing schemas remain valid with the same
+semantics as they had before this RFC.
+
+## Alternatives
+
+Instead of `::id`, we could use many other syntaxes.
+One proposal in the original
+[cedar#579](https://github.com/cedar-policy/cedar/issues/579) thread was to use
+something in the `__cedar` namespace, e.g., `__cedar::schema::top_level::id`.
+Other possibilities include other special characters, like `^id` or `#id`, or even
+a special keyword, like `root id`, which should be unambiguous in both the human
+and JSON schema grammars.
+
+## Unresolved questions
+
+* Should we also support the `::id` syntax in policies?
```

Comment:
At a minimum we'll need to update our parsers so they don't barf on `::id`. People will make this mistake.